### ISSUE TITLE
Add Adobe PDF Accessibility Scan Service

### ISSUE THREAD

**Issue by bbertucc** [bbertucc/equalify#528]
## Problem
Title II laws require accessibility PDFs, but Equalify doesn't have any way to test them. axe does have a [PDF ruleset](https://dequeuniversity.com/rules/axe/pdf/2.0), but it is lacking, when compared to Adobe's own PDF accessibility checking (see [Adobe PDF Accesssibility API Docs](https://developer.adobe.com/document-services/docs/overview/pdf-services-api/howtos/pdf-accessibility-checker-api/). 

## Solution
Integrate Adobe's PDF ruleset to run against PDF documents.

I imagine we'll have to add some way to conditionally run the scan only on PDFs and not on web pages.

We'll also have to figure out how workers treat the long time it takes to return a pdf.

We'll also have to convert the output via https://github.com/EqualifyEverything/equalify-schema

That means we might have to update the schema, and we definitely have to add a new converter.

Deadline is June 1 to deliver report to UIC team, which needs to include a scan of all UIC PDFs run through this system.

## Steps to Close This Issue
- [x] @bbertucc reviews issue with the assignee.
- [x] Assinee budgets issues.
- [x] Budget is approved/disapproved with comment.
- [x] Work commences
- [x] @bbertucc closes this issue after validating fix logging payment.

**Comment by bbertucc** [comment]
@bbertucc is asking Adobe if they have a per page time. That will affect what we test for.

**Comment by bbertucc** [comment]
Adding sample provided by adobe and will ping you with API keys @azdak

**Comment by bbertucc** [comment]
[accessibilityChecker2025-04-04T23-56-29.pdf](https://github.com/user-attachments/files/19613749/accessibilityChecker2025-04-04T23-56-29.pdf)

Adding an example of JSON output from PDF accessibility checker and PDF I checked (took exactly 6 seconds)

[accessibilityChecker2025-04-04T23-56-29.json](https://github.com/user-attachments/files/19613746/accessibilityChecker2025-04-04T23-56-29.json)

**Comment by bbertucc** [comment]
At today's meeting, @azdak was talking about building a separate PDF system. We're dealing with a big amount of PDFs (>200k).

One question: where is the outline of areas that fail accessibility tests?

**Comment by bbertucc** [comment]
Today, we discussed the project with @azdak - timeline is toward the end of May.

**Comment by bbertucc** [comment]
Update: We're not having luck with adobe to get location info. But we're going to proceed with page by page scanning.

@azdak will create a roadmap with estimate on the project.

**Comment by bbertucc** [comment]
@azdak I found additional documentation, still no location info: https://developer.adobe.com/document-services/docs/overview/pdf-services-api/howtos/pdf-accessibility-checker-api/

**Comment by bbertucc** [comment]
Just noting a possible remediation step we could add: https://smartchallenges.asu.edu/challenges/pdf-accessibility-ohio-state-university

Test Site: https://www.remediate-pdf.com/home

That is not something within this ticket scope, but just wanted to post it somewhere for future reference.

**Comment by bbertucc** [comment]
Update: We're getting going on the PDF! The idea is that we're just going with document level issues.

**Comment by bbertucc** [comment]
@azdak here is the official response on pinpointing issues from the Product Manager of the API:

> If I understand correctly, you're talking about errors in terms of Accessibility, returned by the Checker, correct? i.e. not software errors with running the API. Unfortunately, we don't have any immediate solution here as the Checker in its current form is meant more as a report than a tool to precisely identify which parts of a PDF need to be fixed. It's on our long list of improvements as the Checker should also serve in assisting with fixing files automatically, but we're not there yet.

So, it is as we thought.

**Comment by azdak** [comment]
## Adobe PDF Accessibility Checker API Integration Roadmap
#### Data Routing
- [x] **scan**: add new pdf queue
- [ ] **scan/generate**: add new optional flag for enabling PDF detection/scanning
- [x] **scan/generate**: if pdf flag is enabled, catch PDF urls ending in .pdf
  - [x] pass detected PDF urls to new pdf queue
- [x] **scan-worker**: if pdf flag is enabled, catch PDF urls _not_ ending in .pdf
  - [ ] pass detected PDF url to pdf queue 
#### Scanning
- [x] Create new **pdf-scan-worker**; this will allow simple scaling and prevent PDF scans from impacting axe/html scan performance
- [x] pdf queue passes previously detected PDF url to **pdf-scan-worker**
- [x] **pdf-scan-worker**: 
  - [x] downloads PDF locally, and uploads file to adobe API
  - [x] polls adobe API until finished
  - [x] formats results and stores at results/ endpoint
  - [x] deletes stored PDF after successful adobe scan

### Unknowns/Possible Gotchas:
- Due to Adobe costs, system must be opt-in and only enabled if pdf scanning flag is enabled in **/generate** request
- Adobe results will need to be reformatted as either individual Equality issues or as a single, document-level issue (Also need to do this with Editoria11y)
- Headless Puppeteer has a known limitation with opening PDFs, so for urls not ending in .pdf we'll need to add a catch checking the header type(?) in scan-worker
- As with scan, will need to be mindful of handling failed states and logging any issues
- As with scan, much of the work on this feature won't be the MVP, but will come with optimization and dealing with bottlenecks at scale:
  - Depending on the speed of the Adobe API, polling for results might need to be moved to another non-blocking process
  - Saving the PDF locally in the pdf-worker may not be ideal; Adobe also supports input files uploaded to signed storage (eg amazon s3)

**Estimate for MVP**: 1-2 weeks billed

**Comment by bbertucc** [comment]
We'll test it in the dev demo page. 

I want to flag that we will be doing multiple PDFS.

**Comment by bbertucc** [comment]
Update: This week he'll be scaffolding the service.

**Comment by azdak** [comment]
Updated punchlist for API integration:
- [x] The new version slightly changed the way jobIds are generated in scan; instead of an ascending integer, IDs are now a string: "eq-"+[ascending integer]. Everything else remains the same, but this may require adjustment if the API relies on jobIds as ints when consuming results. (cc: @heythisischris)
- [x] PDF urls _not_ ending in .pdf are not yet caught by the system, and will introduce some complexity. This could be handled a couple of ways: 1) entirely on the scan side (complicated) 2) return an error/warning, and add a "forcePDF" option when (re-)submitting urls (definitely simpler, imo, but deserves a discussion with @heythisischris)
- [x] As mentioned in Unknowns, Adobe results (and Editoria11y results) will need to be reformatted as either individual Equality issues or as a single, document-level issue; really just need to sit down with it and figure out the best way to handle it.
- [ ] Unclear on how we want to enable the checks for PDFs; could be a simple "enablePDFScan" type flag when submitting to scan, or could be something more secure/complicated and tied to accounts
- [x] Update [the equalify demo](https://github.com/EqualifyEverything/scan-demo.equalify.app) to integrate new PDF capabilities

**Comment by bbertucc** [comment]
A few new tasks:
- [x] `generate` endpoints: New parameter - API key (salted)
- [ ] `generate` endpoints: New parameter - Enable PDF scan
- [x] `generate` endpoints: New parameter - Force PDF scan
- [x] Standard scan worker: Puppeteer: Add catch that looks at MIME type to handle PDFs, return "this URL is a PDF" status

**Comment by bbertucc** [comment]
This has been deployed, ready for the upcoming released
